test/e2e/knit_irqwatch_test.go

```diff
@@ -0,0 +1,227 @@
+
+			ok := areCountersEqual(delta, counters)
+			o.Expect(ok).To(o.BeTrue())
+		})
+	})
+
+	g.BeforeEach(func() {
+		snapshotRoot = snapshotBeforeEach(fixtureName, "sysinfo.tgz")
+	})
+
+	g.AfterEach(func() {
+		snapshotAfterEach(snapshotRoot)
+	})
+})
+
+func modifyInterruptsFile(procFs string) (irqs.Stats, error) {
+	mockLog := log.New(ioutil.Discard, "", 0)
+
+	ih := irqs.New(mockLog, procFs)
+	stats, err := ih.ReadStats()
+	if err != nil {
+		return nil, err
+	}
+	initialStats := stats.Clone()
+	delta := make(irqs.Stats)
+	rand.Seed(time.Now().UnixNano())
+
+	// 50 random changes are good enough
+	const numOfModifications int = 50
+	// just a number that is big enough to make nice distribution of interrupts on random queues
+	const maxInterrupts = 1000
+
+	for i := 0; i < numOfModifications; i++ {
+		numOfCpus := len(stats)
+
+		// select random cpuid to change its counters
+		cpuid := int(rand.Intn(numOfCpus))
+
+		// select random IRQ in the seleced cpuid
+		k := selRandKey(stats[cpuid])
+
+		stats[cpuid][k] += uint64(rand.Intn(maxInterrupts))
+		delta[cpuid] = make(irqs.Counter)
+		delta[cpuid][k] = stats[cpuid][k] - initialStats[cpuid][k]
+	}
+
+	// add the random new stat values to the interruts file
+	return delta, reWriteInterrupts(procFs, stats)
+}
+
+func reWriteInterrupts(procFs string, stats irqs.Stats) error {
+	tmpInterrruptsFile := filepath.Join(procFs, "interrupts_temp")
+	interrruptsFile := filepath.Join(procFs, "interrupts")
+
+	f, err := os.Create(tmpInterrruptsFile)
+	if err != nil {
+		g.Fail(fmt.Sprintf("fail to create temp interrupts file %q", tmpInterrruptsFile))
+		return err
+	}
+
+	defer f.Close()
+
+	// Since it's a test we can assume no offlined cpu, so no holes in the slice (i.e. cpu_0 to cpu_n are presents)
+	type cpuidToIrqValue []uint64
+
+	// IRQ name maps to counter values for each cpuid
+	// this way is makes it easier to be wrriten back into a file, line by line
+	irqMap := make(map[string]cpuidToIrqValue)
+
+	cpuids := getSortedCPUids(stats)
+
+	for cpuid := range cpuids {
+		for k, v := range stats[cpuid] {
+			irqMap[k] = append(irqMap[k], v)
+		}
+	}
+
+	tab(f)
+	// write cpu ids on the first line
+	for cpuid := range cpuids {
+		cpuName := fmt.Sprintf("%11s", fmt.Sprintf("CPU%d", cpuid))
+		f.WriteString(cpuName)
+	}
+	newLine(f)
+
+	for irqName, counters := range irqMap {
+		f.WriteString(fmt.Sprintf("%4s:", irqName))
+
+		for _, v := range counters {
+			f.WriteString(fmt.Sprintf("%11d", v))
+		}
+		newLine(f)
+	}
+
+	// use this approach in order to avoid from other processes to access
+	// the original interrupts file while modifing it
+	os.Rename(tmpInterrruptsFile, interrruptsFile)
+	if err != nil {
+		g.Fail(fmt.Sprintf("fail to rename file %q to %q", tmpInterrruptsFile, interrruptsFile))
+		return err
+	}
+
+	return nil
+}
+
+func tab(f *os.File) {
+	f.WriteString("\t")
+}
+
+func newLine(f *os.File) {
+	f.WriteString("\n")
+}
+
+func selRandKey(c irqs.Counter) string {
```

Comment:
I'd use something along the lines of https://stackoverflow.com/questions/33994677/pick-a-random-value-from-a-go-slice
1. make a slice of keys
2. pick a random key using `rand.Intn`